src/renderer/uia/UiaRenderer.cpp

```diff
@@ -223,6 +234,31 @@ CATCH_RETURN();
 // Return Value:
 // - S_OK, else an appropriate HRESULT for failing to allocate or write.
 [[nodiscard]] HRESULT UiaEngine::EndPaint() noexcept
+{
+    // Snap this now while we're still under lock
+    // so present can work on the copy while another
+    // thread might start filling the next "frame"
+    // worth of text data.
+    _queuedOutput = std::move(_newOutput);

```

Comment:
You can make your engine reuse previously allocated memory, by writing:
```cpp
std::swap(_queuedOutput, _newOutput);
_newOutput.clear();
```

This works because `BeginPaint`, `EndPaint`, etc. aren't called again until the previous `Present()` has finished.

Once you do that the `_newOutput += newText;` will be very performant since memory copies are extremely fast (in the order of >10GB/s).